### ISSUE TITLE
[Feature]Streamline HVPA Status

### ISSUE THREAD

**Issue by amshuman-kr** [gardener/hvpa-controller#14]
**Feature (What you would like to be added):**
At present, the `HvpaStatus` has three sub-structures.
```
// HvpaStatus defines the observed state of Hvpa
type HvpaStatus struct {
	HpaStatus HpaStatus `json:"hpaStatus,omitempty" protobuf:"bytes,1,opt,name=hpaStatus"`
	VpaStatus vpa_api.VerticalPodAutoscalerStatus `json:"vpaStatus,omitempty" protobuf:"bytes,2,opt,name=vpaStatus"`
	HvpaStatus HvpaCurrentStatus `json:"hvpaCurrentStatus,omitempty" protobuf:"bytes,3,opt,name=hvpaCurrentStatus"`
}
```
   - `HpaStatus` and `VpaStatus` are continuously updated based on recommendations from the HPA and VPA status correspondingly.
   - `HvpaCurrentStatus` has the actual status of HVPA as to when the HVPA actually applies the HPA and VPA recommendations based on the various stabilization criteria including stabilization duration, [maintenance window](https://github.com/gardener/hvpa-controller/issues/12 and the [override](https://github.com/gardener/hvpa-controller/blob/master/pkg/apis/autoscaling/v1alpha1/hvpa_types.go#L152) of stabilization.

We should restructure the status to convey all and only the information relevant about HVPA such as..
1. When HVPA scaled (or would have liked to scale of `UpdatePolicy` allowed)
1. Which direction HVPA scaled (or would have liked to scale of `UpdatePolicy` allowed)
1. Which resource HVPA scaled (or would have liked to scale of `UpdatePolicy` allowed)
1. To what value HVPA scaled (or would have liked to scale of `UpdatePolicy` allowed)

**Motivation (Why is this needed?):**
`HvpaStatus` should be the single source of truth about the behaviour (or recommended behaviour if `UpdatePolicy` does not allow) of HVPA. If someone is interested in HPA/VPA recommendations, they can always to directly to the HPA/VPA resources.

**Approach/Hint to the implement solution (optional):**
A possible structure of `HvpaStatus` could be as follows.
```
// HvpaStatus defines the observed state of Hvpa
type HvpaStatus struct {
    updatePolicy: UpdatePolicy `json:"updatePolicy"`
    HpaWeight VpaWeight `json:"hpaWeight,omitempty"`
    VpaWeight VpaWeight `json:"vpaWeight,omitempty"`
    OverrideScaleUpStabilization bool `json:"overrideScaleUpStabilization,omitempty"`
    LastScaling map[ScalingCategory"]ScalingStatus `json:"lastScaling,omitempty"`
    LastOperation ...
    LastError ...
}

type UpdatePolicy struct {
	UpdateMode *UpdateMode `json:"updateMode,omitempty" protobuf:"bytes,1,opt,name=updateMode"`
}

type UpdateMode string

const (
	UpdateModeOff UpdateMode = "Off"
	UpdateModeScaleUp UpdateMode = "ScaleUp"
	UpdateModeOn UpdateMode = "On"
)

type ScalingCategory string

const (
	ScalingCategoryApplied ScalingCategory = "Applied"
	ScalingCategoryBlockedByUpdatePolicy ScalingCategory = "BlockedByUpdatePolicy"
	ScalingCategoryBlockedBySpec ScalingCategory = "BlockedBySpec"
)

type ScalingStatus struct {
    LastScaleTime *metav1.Time `json:"lastScaleTime,omitempty"`
    LastScaleType LastScaleType `json:"lastScaleType,omitempty"`
    HpaStatus HpaStatus `json:"hpaStatus,omitempty" protobuf:"bytes,1,opt,name=hpaStatus"`
    VpaStatus vpa_api.VerticalPodAutoscalerStatus `json:"vpaStatus,omitempty" protobuf:"bytes,2,opt,name=vpaStatus"`
}
```

**Comment by amshuman-kr** [comment]
How about something like this ...?

```go
// HvpaStatus defines the observed state of Hvpa
type HvpaStatus struct {
    updatePolicy: UpdatePolicy `json:"updatePolicy"`
    HpaWeight VpaWeight `json:"hpaWeight,omitempty"`
    VpaWeight VpaWeight `json:"vpaWeight,omitempty"`
    OverrideScaleUpStabilization bool `json:"overrideScaleUpStabilization,omitempty"`
    LastBlockedScaling []BlockedScaling
    LastScaling ScalingStatus `json:"lastScaling,omitempty"`
    LastOperation ...
    LastError ...
}

type UpdatePolicy struct {
	UpdateMode *UpdateMode `json:"updateMode,omitempty" protobuf:"bytes,1,opt,name=updateMode"`
}

type UpdateMode string

const (
	UpdateModeOff UpdateMode = "Off"
	UpdateModeScaleUp UpdateMode = "ScaleUp"
	UpdateModeOn UpdateMode = "On"
)

type BlockingReason string

const (
	BlockingReasonStabilizationWindow BlockingReason = "StabilizationWindow"
	BlockingReasonMaintenanceWindow BlockingReason = "MaintenanceWindow"
	BlockingReasonUpdatePolicy BlockingReason = "UpdatePolicy"
	BlockingReasonWeight BlockingReason = "Weight"
	BlockingReasonMinChange BlockingReason = "MinChange"
)

type ScalingStatus struct {
    LastScaleTime *metav1.Time `json:"lastScaleTime,omitempty"`
    LastScaleType LastScaleType `json:"lastScaleType,omitempty"`
    HpaStatus HpaStatus `json:"hpaStatus,omitempty" protobuf:"bytes,1,opt,name=hpaStatus"`
    VpaStatus vpa_api.VerticalPodAutoscalerStatus `json:"vpaStatus,omitempty" protobuf:"bytes,2,opt,name=vpaStatus"`
}

type BlockedScaling struct {
    ScalingStatus ScalingStatus `json:"scaling,omitempty"
    Reason BlockingReason `json:"reason,omitempty"
}
```

**Comment by ggaurav10** [comment]
There can be cases where VPA recommendations are partially applied/blocked. For eg - memory can be applied because it passes all the requirements, whereas CPU recommendation may be blocked because of `MinChange`. In this case, we will be duplicating the VPA status under both:
```
LastBlockedScaling []BlockedScaling
LastScaling ScalingStatus `json:"lastScaling,omitempty"
```

LastBlockedScaling will have the CPU recommendation only whereas LastScaling will have the memory recommendations. Right?

To make matters worse, multiple containers in a pod can have scaling blocked because of different reasons. A common field in the spec may not capture all. I think we need to extend the spec some more so that it doesn't reflect inaccurate status. WDYT?

**Comment by amshuman-kr** [comment]
> In this case, we will be duplicating the VPA status under both:
> ...
> To make matters worse, multiple containers in a pod can have scaling blocked because of different reasons.

Technically, we won't be duplicating. Only sorting.

> A common field in the spec may not capture all. I think we need to extend the spec some more so that it doesn't reflect inaccurate status. WDYT?
I proposed `BlockingReason` to mitigate this exact problem. But why spec? Only status is enough, no? Or are you thinking about different minChange for each container?

**Comment by amshuman-kr** [comment]
Also, if more than one reason blocks the same change to the same container, we can sort it according to our convenience. WDYT? Otherwise we can't optimise based on fail-fast.

**Comment by amshuman-kr** [comment]
I propose this order of preference for blocking, Weight>UpdatePolicy>StabilizationWindow/Mainten>MinChanged. WDYT?

**Comment by amshuman-kr** [comment]
But any other fixed order is also fine. Whatever that makes fail fast easier.

**Comment by amshuman-kr** [comment]
We might need a `State` field in the HvpaStatus also to indicate if the last scaling was successfully applied or we need to retry

**Comment by ggaurav10** [comment]
> I propose this order of preference for blocking, Weight>UpdatePolicy>StabilizationWindow/MaintenanceWindow>MinChanged. WDYT?

lgtm

**Comment by amshuman-kr** [comment]
Another iteration (added state and status subresource fields). The [status subresource](https://book-v1.book.kubebuilder.io/basics/status_subresource.html) annotation should be added. Regarding `Scale` subresource annotations please refer [here](https://github.com/kubernetes-sigs/kubebuilder/issues/554).

Questions:
- Will we need [`CollisionCount`](https://github.com/kubernetes/api/blob/0772a1bdf941/apps/v1/types.go#L412)?
- Is it ok to point 

```go
// HvpaStatus defines the observed state of Hvpa
type HvpaStatus struct {
	ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"varint,1,opt,name=observedGeneration"`
	State HvpaState `json:"state"`
	Replicas *int32 `json:"replicas,omitempty" protobuf:"varint,1,opt,name=replicas"` // For Scale.Status.Replicas, to work in the future as a targetRef for HPA
	TargetPodSelector *metav1.LabelSelector `json:"selector" protobuf:"bytes,2,opt,name=selector"` // Derived from the targetRef's Scale.spec.selector  to work in the future as a targetRef for HPA
	HpaWeight VpaWeight `json:"hpaWeight,omitempty"`
	VpaWeight VpaWeight `json:"vpaWeight,omitempty"`
	OverrideScaleUpStabilization bool `json:"overrideScaleUpStabilization,omitempty"`
	LastBlockedScaling []BlockedScaling
	LastScaling ScalingStatus `json:"lastScaling,omitempty"`
	LastOperation ...
	LastError ...
}

type HvpaState string

const (
	HvpaStateReconciled HvpaState = "Reconciled"
	HvpaStateReconciled HvpaState = "Pending"
)

type BlockingReason string

const (
	BlockingReasonStabilizationWindow BlockingReason = "StabilizationWindow"
	BlockingReasonMaintenanceWindow BlockingReason = "MaintenanceWindow"
	BlockingReasonUpdatePolicy BlockingReason = "UpdatePolicy"
	BlockingReasonWeight BlockingReason = "Weight"
	BlockingReasonMinChange BlockingReason = "MinChange"
)

type ScalingStatus struct {
	LastScaleTime *metav1.Time `json:"lastScaleTime,omitempty"`
	LastScaleType LastScaleType `json:"lastScaleType,omitempty"`
	HpaStatus HpaStatus `json:"hpaStatus,omitempty" protobuf:"bytes,1,opt,name=hpaStatus"`
	VpaStatus vpa_api.VerticalPodAutoscalerStatus `json:"vpaStatus,omitempty" protobuf:"bytes,2,opt,name=vpaStatus"`
}

type BlockedScaling struct {
	Reason BlockingReason `json:"reason,omitempty"
	ScalingStatus
}
```

Let's also enhance `HvpaSpec` to include `Replicas`.
```go
type HvpaSpec struct {
	...
	Replicas *int32 `json:"replicas,omitempty" protobuf:"varint,1,opt,name=replicas"` // For Scale.Spec.Replicas, to work in the future as a targetRef for HPA
	...
```

Also, let's introduce the `Purge` enumeration value for both HPA and VPA update policy as mentioned [here](https://github.com/gardener/hvpa-controller/issues/13#issuecomment-522532848).

**Comment by ggaurav10** [comment]
Fixed in #19
